source/reference/best-practices.rst

```diff
@@ -28,15 +28,60 @@ For an AppImage to run on most systems, the following conditions need to be met:
 Binaries must not use compiled-in absolute paths
 ------------------------------------------------
 
-Since an AppImage is mounted at a different location in the filesystem every time it is run, it is crucial not to use compiled in absolute paths. For example, if the application accesses a resource such as an image, it should do so from a location relative to the main executable. Unfortunately, many applications have absolute paths compiled in (:code:`$PREFIX`, most commonly :code:`/usr`) at compile time.
-
+Since an AppImage is mounted at a different location in the filesystem
+every time it is run, it is crucial not to use compiled in absolute
+paths. For example, if the application accesses a resource such as an
+image, it should do so from a location relative to the main
+executable. Unfortunately, many applications have absolute paths
+compiled in (:code:`$PREFIX`, most commonly :code:`/usr`) at compile
+time.
+
+The canonical way on Linux to construct a relative path is to first
+resolve ``proc/self/exe``. That provides the path to the main
+executable. As a result, it should work both in normal installations
+and in relocatable installations such as AppImages.
 
 .. _ref-open-source-applications:
 
 Open source applications
 ^^^^^^^^^^^^^^^^^^^^^^^^
 
-Wherever possible you should change the Source Code of the application in order not to use absolute paths. There are several ways to do this. The canonical way on Linux is to resolve ``proc/self/exe`` to get the path to the main executable and construct a relative path from there. As a result, it should work both in normal installations and in relocatable installations such as AppImages.
+You can check for hard-coded absolute paths:
+
+.. code-block:: shell
+
+	strings MyApp.AppDir/usr/bin/myapp | grep /usr
+
+Should this return something, then you need to modify your app programmatically (e.g., by using relative paths, using `binreloc <https://github.com/limbahq/binreloc>`__, or using :code:`QString QCoreApplication::applicationDirPath()`).
```

Comment:
I think part of this PR should include some improvement here. The link to the binreloc repo works, but in its README, the link to the docs on how to actually use it is dead. Someone reported it a while ago but it hasn't been corrected. https://github.com/limbahq/binreloc/issues/4 I've never used binreloc. Is it still stable? Is there a good site that has the docs on how to use it?

Reply:
It's just a library that helps developers use paths relative to the current binary. Many GUI frameworks also include such facilities, a famous example is Qt's [`QCoreApplication::applicationDirPath()` method](https://doc.qt.io/qt-6/qcoreapplication.html#applicationDirPath). 

Generally this section is quite outdated. We should include more real-world examples from common libs like Qt, Gtk+. The parent section already explains that resources must be located relative to the main binary (arguably the wording could be improved).

Reply:
This is mostly done I think.

In places where I edited or added new content, I wrapped long lines because it's much easier for me to review and edit documents. In fact, many projects I've worked on have guidelines that state long lines should be wrapped around col 79 (or something similar). I hope that will be fine for the purposes of this PR.

> Generally this section is quite outdated. We should include more real-world examples from common libs like Qt, Gtk+.

Any ones in particular you'd like added?

You'll see I removed all references to binreloc. As the binreloc project repo (that was linked in the docs here) wasn't updated since 2016, I couldn't find the API docs for it (just a lot of dead links), and it doesn't seem to be included in any LInux repos https://repology.org/project/binreloc/versions

Reply:
For Gtk, the recommended approach would be to compile any external resources _into_ the application wherever possible, as [GResource](https://docs.gtk.org/gio/struct.Resource.html)s.

You create a manifest file as a `.gresource.xml` file, and the `glib-compile-resources` tool can either compile it to a `.gresource` binary file, or (with `--generate-source` and `--generate-header`) it can output C code to compile into the application.

Whether resources come from a `.resource` file loaded with   `g_resource_load()` or are directly compiled in and registered by calling `register_resource()`, the contents can be accessed using `g_resource_open_stream()` / `g_resources_lookup_data()`, or by instantiating a `GFile` to a resource URL. (`resource:///org/gtk/Example/data/splashscreen.png` is the example their docs use.)

I don't believe Gtk offers any particular features for determining or using the application runtime location; their filesystem tools are mostly focused on things like standard XDG paths.

(It might be worth mentioning, too, that Qt also has the ability to use compiled-in resources: a `.qrc` file acts as the manifest, and their `rcc` compiler takes the place of `glib-compile-resources` in generating code for incorporation into the application itself. `QFile`, and any other Qt class that accepts file path strings, can be passed a string of the form `":/path/to/resource"` to load a compiled-in resource instead. 

The only exception is PyQt6, which dropped support for the Qt resource system. PyQt5 included a `pyrcc5` which compiled resources into Python code, and Qt's `rcc` compiler now supports compiling resources into Python for use with Qt for Python 6.)

Reply:
(I wrote that like Qt was aping GLib, but credit where credit is due: AFAIK Qt's resource system _long_ preceded GResource, and was very likely a big part of the inspiration for its design. GResource didn't exist until Gio 2.32 in 2012, but Qt has had a resource system since Qt 4 (2005). There was even a much more limited predecessor system called `qembed` in Qt 3.)